### ISSUE TITLE
Bug: @typescript-eslint/no-use-before-define `"typedefs": true` should fail

### ISSUE THREAD

**Issue by mojito317** [typescript-eslint/typescript-eslint#10389]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=DYUwLgBAtgngagQwE4C4IGUxIJYDsDmA8kgHICuUARiEgNwBQYMADiBlnkaRdUhALwQAzhwIQAPhFw8atIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRQgI0QDN9oyATagS10XSnyqsgBpxtIAAQAuATwAOKAMbQm44YRTwWwgPQFipCtVoNmrdgG1%2B2KImjRavE9iymBYyYyqoMwuIhsQAvjYC6-L7eQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
let myVar: StringOrNumber;
type StringOrNumber = string | number;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": [
      "error",
      {
        "typedefs": true
      }
    ]
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect a failure, as the reference to the type is before the type declaration.

### Actual Result

No failure.

### Additional Info

If I understand [the documentation](https://typescript-eslint.io/rules/no-use-before-define/#typedefs) correctly, `{no-use-before-define: "off", @typescript-eslint/no-use-before-define: "error"}` should fail if a type has a reference before declaration.

**Comment by mojito317** [comment]
Also: this was an issue (#249) once before, but closed as resolved.

**Comment by kirkwaiblinger** [comment]
Hi @mojito317 !

Yeah, that seems like a clear bug. In fact, I couldn't get any tests in our repo to fail by forcing the typedefs option to be true/false.... So that option clearly needs some attention!

**Comment by bradzacher** [comment]
Sounds like we need to remove one of the options because they are conflicting?

**Comment by developer-bandi** [comment]
@bradzacher I think so. If I had to be specific, I think it would be better to remove the ignoreTypeReferences option because the typedefs option seems a bit clearer.

However, there must have been a good reason for the ignoreTypeReferences option to be created, so it seems like it would be better to first understand the history.

**Comment by StyleShit** [comment]
IIUC, these 2 options are doing basically the same, just with different defaults.

Currently, `ignoreTypeReferences` takes precedence, so (as said above already) only if you set `ignoreTypeReferences: false`, you can turn on `typedefs`. Which should have been on by default, but it actually can't due to `ignoreTypeReferences` being `true` by default...

So yes, these are 2 completely conflicting options, we need to remove one of them, and I agree with @developer-bandi that it should be `ignoreTypeReferences`.

It's probably a breaking change, but considering this unique case, I think it could be fairly safe to do it, assuming we'll keep the default behavior (e.g., `typedefs` will need to be `false` by default).

Does it fall under "[Fixes incorrect behavior in a rule that may or may not introduce additional reports.](https://typescript-eslint.io/users/versioning/#eslint-plugin:~:text=Fixes%20incorrect%20behavior%20in%20a%20rule%20that%20may%20or%20may%20not%20introduce%20additional%20reports)"?


Pinging @typescript-eslint/triage-team for opinions
